Moves random source generation into helper library

TyOverby opened the pull request:
Closes #7159

TyOverby commented:
@gafter @dotnet/roslyn-compiler

jmarolf commented on src/Compilers/VisualBasic/Test/Syntax/BasicCompilerSyntaxTest.vbproj:
```diff
@@ -96,7 +96,7 @@
     <Compile Include="LocationTests.vb" />
     <Compile Include="Parser\DeclarationTests.vb" />
-    <Compile Include="Parser\FuzzTesting.vb" />
+    <Compile Include="Parser\ParserRegressionTests.vb" />
```
:+1:

jaredpar commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,28 @@
+{
+    public static class SourceUtilities
+    {
+        public static String GenerateLargeRandomString()
```
Use `string` instead of `String`

TyOverby replied:
Fixed

jaredpar commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,28 @@
+﻿using System;
```
Need a copyright header

TyOverby replied:
Done

davkean commented:
Linux failure due to #7411, I've taken the machine offline.

@dotnet-bot retest prtest/lin/dbg/unit32 please

AlekseyTs commented:
LGTM

jaredpar commented:
:+1:

TyOverby commented:
@gafter wants me to remove this test entirely (along with the corresponding C# one).  Apparently his machine fails the C# version of this test (with the fix!) one out of ten times, making it a huge hinderance to getting work done.

jaredpar commented:
We should focus on making this valuable test more reliable vs. deleting it.  

Why is this test failing on his machine.  Is it an OOM?  If so I'm guessing it's because it requires a contiguous allocation.  Have we considered breaking it up into chunks, deriving from `SourceText` and driving the parser that way?

TyOverby commented:
@jaredpar Yeah, it's failing on OOM. 

> deriving from SourceText

That's a good idea.  I'll play with an implementation.

TyOverby commented:
@jaredpar @gafter:

This should fix the flakiness.  

I did what Jared suggested and subclassed SourceText to be a list of smaller buffers.

TyOverby commented:
@dotnet/roslyn-compiler 

I made a pretty big change since the last reviews.  If I could get someone to look over this code again, that would be great!

AlekseyTs commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,90 @@
+                    i = 0;
+                    blocks.Add(new char[BLOCK_SIZE]);
+                }
+                blocks.Last()[i] = ch;
```
Consider capturing in a local the current block that we are filling instead of calling blocks.Last() for each character.

AlekseyTs commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,90 @@
+            {
+                if (blocks.Count != 0)
+                {
+                    return (blocks.Count - 1) * BLOCK_SIZE + blocks.Last().Length;
```
Is there an expectation that ```blocks.Last().Length``` could be different from BLOCK_SIZE?

AlekseyTs commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,90 @@
+            {
+                var block = position / BLOCK_SIZE;
+                var offset = position % BLOCK_SIZE;
+                return blocks[block][offset];
```
It looks like an assumption is made that the last block is completely filled with data? What this assumption is based on?

AlekseyTs commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,90 @@
+
+        public static SourceText GenerateLargeRandomString()
+        {
+            return new SegmentedSourceText(RandomString());
```
It looks like SegmentedSourceText keeps all the characters in memory, can this be avoided? For example, if the tests that are using this SourceText never access characters in random order, but always do this consecutively, we could have implementation of SourceText that doesn't cache any characters in memory, but rather "generates" them on demand.

TyOverby replied:
I'll look into it.

gafter commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,50 @@
+        private char hash(int position)
+        {
+            // good enough "random looking" hasher via the xorshift128plus algorithm.
+            UInt64 state0 = (UInt64) 1;
```
This does not appear to be related to the data previously processed by the test.

TyOverby replied:
The specific random data from the previous test was mostly irrelevant.  The only thing that mattered was generating more syntax errors than we could handle.

TyOverby commented:
@AlekseyTs: This PR now streams random characters into the lexer and parser.  No more allocations at all!

gafter commented on src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb:
```diff
@@ -918,20 +919,8 @@ End Enum
-        Dim tree = VisualBasicSyntaxTree.ParseText(source)
-
-        Assert.Equal(source, tree.ToString())
+        Dim tree = VisualBasicSyntaxTree.ParseText(New RandomizedSourceText())
```
Do we have some reason to believe that this modified test would have triggered the same (original) issue?

gafter replied:
OK, I'm willing to believe it would.

TyOverby replied:
The original issue was "we try to read binary files, and we generate millions of errors to report back to the user; this crashes us."  

Although this doesn't use the same seed, it will trigger the original issue.

AlekseyTs replied:
This sounds reasonable to me.

gafter commented on src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs:
```diff
@@ -83,24 +81,7 @@ public void c024928()
-            }
-
-            var tree = CSharpSyntaxTree.ParseText(str);
+            CSharpSyntaxTree.ParseText(new RandomizedSourceText());
```
Do we have some reason to believe that this modified test would have triggered the same (original) issue?

The (flakiness) failure I frequently see in this test is in the debug code inside the parser written to verify that the parsed text is consistent with the original; it creates a 40MB string to do that, which exhausts memory when tests are run in parallel. This doesn't really change that.

TyOverby replied:
This code gets to the parsing phase?  I thought for sure that the errors that we generate come before we have a parse tree to validate.

gafter replied:
Most of the failures I've seen recently have been in the parser where it validates that the input == the syntax tree stringified.

TyOverby replied:
Really?  How do we get a parse tree if the text is literally unparsable?  What parse tree would be generated by the text "if if if if if if if if \0"?

AlekseyTs replied:
We are producing parse trees for any text given as an input. Even when there are parsing/scaning errors, the parse tree is supposed produce the original text when converted to string.

TyOverby commented:
test this please

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs:
```diff
@@ -83,24 +81,7 @@ public void c024928()
         [Fact]
         public void TestBinary()
```
Note, there is another PR that disables the test make sure the test is enabled once this change is merged.

TyOverby replied:
Yep, I'll pay attention during merge.  Fortunately Git should flag it as a conflict.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Syntax/Parser/ParserRegressionTests.vb:
```diff
@@ -918,20 +919,8 @@ End Enum
 
     <WorkItem(2867, "https://github.com/dotnet/roslyn/issues/2867")>
-    <Fact(Skip:="https://github.com/dotnet/roslyn/issues/7159")>
     Public Sub TestBinary()
```
The ```Fact``` attribute must be preserved without the Skip reason.

TyOverby replied:
Done.

AlekseyTs commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,50 @@
+            {
+                var s1 = state0;
+                var s0 = state1;
+                state0 = s0;
```
It looks like ```state0``` is not used after this point. Is there a reason to change its value? Can it be a const local?

AlekseyTs commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,50 @@
+                s1 ^= s1 >> 17;
+                s1 ^= s0;
+                s1 ^= s0 >> 26;
+                state1 = s1;
```
Is there a particular reason why we need ```state1``` local? It looks like we can simply initialize ```s0``` with ```position``` and then use ```s1``` in place of ```state1``` in the return statement. If I am not missing something, the code will be clearer.

TyOverby replied:
The "reason" is that I naively copied pseudocode from a wikipedia article.

I've cleaned up the locals and also changed state0 and state1 to be better chosen "constants".

cston replied:
Can we avoid this approach and do something simpler? Perhaps ```return builder[position % builder.Count()]``` where `builder` is 1K version of the buffer we were creating originally.

AlekseyTs commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,50 @@
+            }
+        }
+
+        public override char this[int position] => hash(position);
```
Should we throw on index out of range?

TyOverby replied:
We could?  It would slow things down and not provide any extra test coverage though.

AlekseyTs commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,50 @@
+        {
+            for (var i = 0; i < count; i++)
+            {
+                destination[destinationIndex + i] = hash(sourceIndex);
```
It looks like we fill buffer with the same character, ```sourceIndex``` is never incremented. Also, should we throw throw on an attempt to get characters out of range?

TyOverby replied:
> `sourceIndex` is never incremented

Fixed, thanks.

> Also, should we throw throw on an attempt to get characters out of range?

I doubt this would improve test coverage of that case, but adding it would be easy enough.

gafter replied:
Can you please push your changes into the PR?

TyOverby replied:
Yep; sorry.  Got busy and forgot about this.

AlekseyTs commented on src/Test/Utilities/Shared/Syntax/SourceUtilities.cs:
```diff
@@ -0,0 +1,47 @@
+
+        public override void CopyTo(int sourceIndex, char[] destination, int destinationIndex, int count)
+        {
+            throw new UnauthorizedAccessException();
```
Is implementation of this method not needed for the test? Then we should probably throw NotImplementedException or NotSupportedException and get rid of the remaining code.

TyOverby replied:
Oh, I threw that in there to see if `CopyTo` was being used.  Turns out it was (as you can see in the test failures).

When @gafter asked me to push my fixes to other issues, I forgot to change this back to it's actual implementation.

TyOverby commented:
@gafter @AlekseyTs 

It seems like the only remaining problem is the

> debug code inside the parser written to verify that the parsed text is consistent with the original; it creates two 40MB strings to do that, one for the input and one for the parse tree. -- @gafter 

Do you think that we could have these tests run only on release builds?  This would keep testing for this scenario, but wouldn't do that crazy allocation.  (in fact, this would better test the real world use-case that we originally created this test for)

AlekseyTs commented:
I think running on release only is a reasonable trade-off.

TyOverby commented:
@dotnet/roslyn-infrastructure 

Where are all of these commits coming from?  Did someone rewrite history or did I mess things up?

TyOverby commented:
run prtest/win/dbg/eta please

TyOverby commented:
test prtest/win/dbg/eta please

AlekseyTs commented on src/Test/Utilities/Shared/Assert/ConditionalFactAttribute.cs:
```diff
@@ -49,4 +49,15 @@ public class IsEnglishLocal : ExecutionCondition
+#if DEBUG
+        public override bool ShouldSkip => true;
+#else
+        public override bool ShouldSkip => true;
```
Should this return true?

AlekseyTs replied:
I mean false?

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs:
```diff
@@ -80,27 +78,10 @@ public void c024928()
 
         [WorkItem(2771, "https://github.com/dotnet/roslyn/issues/2771")]
-        [Fact]
+        [ConditionalFact(typeof(IsRelease))]
```
Wouldn't it be simpler to surround this test with ```#if !DEBUG```?

jaredpar commented on src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs:
```diff
@@ -79,28 +77,13 @@ public void c024928()
             var tree = SyntaxFactory.ParseSyntaxTree(test);
         }
 
+#if !DEBUG 
```
Rather than use a conditional directive here, can we introduce `FactReleaseOnly`?  A custom `Fact` type is generally how we control tests that execute under specific conditions.

TyOverby replied:
Ack! that's what I had and Aleksy recommended that I remove it in favor of this.  I can revert and fix the mistake that I had in that version.

TyOverby replied:
Also, we must have "release-only" tests and "debug-only" tests, but I couldn't find any with a quick `git grep`.

AlekseyTs replied:
I suggested to use conditional compilation because the custom fact would also be implemented using conditional compilation, but in a different project. It felt natural to me to simply \#if the unit-test rather than build whole infrastructure to accomplish the same thing.

jaredpar replied:
My reasoning for wanting to use a custom `Fact` is the following:

1. In general I dislike `#if` directives and want to minimize them because it makes it harder to accurately analyze source code as it creates multiple views of the code.  This is a frequent problem in tools like `codeformatter`.  If we run this tool with `Release` defined (which is a valid idea) then it will never process any of this added code.  
2. Our pattern thus far has been to conditionally execute tests by using a custom `Fact` attribute.  Using `#if` manually here goes against that pattern.

gafter commented:
LGTM, though I expect you'll want to address @jaredpar 's comment about using a custom `Fact` attribute for release-only.

TyOverby commented:
Alright, I moved back to using a custom Fact attribute.  If all these tests pass, I'll rewrite the commit message (I don't know how this happened!) and merge.

TyOverby commented:
@jaredpar Could I get one more pass over this before I merge?

jaredpar commented:
:+1:
